lightning/src/util/persist.rs

```diff
@@ -243,10 +250,889 @@ where
+/// ... and so on.
+///
+/// # Reading channel state from storage
+///
+/// Channel state is reconstructed by calling [`Self::read_channel_monitors_with_updates`]. Please
+/// see that function's documentation for more information.
+///
+/// # Pruning stale channel updates
+///
+/// Stale updates are pruned when a full monitor is written. The old monitor is first read, and if
+/// that succeeds, updates in the range between the old and new monitors are deleted. The `lazy`
+/// flag is used on the [`KVStore::remove`] method, so there are no guarantees that the deletions
+/// will complete. However, stale updates are not a problem for data integrity, since updates are
+/// only read that are higher than the stored [`ChannelMonitor`].
+///
+/// To keep storage usage in check, the [`Self::cleanup_stale_updates`] function as needed.
+pub struct MonitorUpdatingPersister<K: Deref, L: Deref, ES: Deref + Clone, SP: Deref + Clone>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	kv_store: K,
+	logger: L,
+	maximum_pending_updates: u64,
+	entropy_source: ES,
+	signer_provider: SP,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref, ES: Deref + Clone, SP: Deref + Clone>
+	MonitorUpdatingPersister<K, L, ES, SP>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	/// Construct a new `MonitorUpdatingPersister`.
+	///
+	/// The `maximum_pending_updates` parameter controls how many updates may be stored before a MUP
+	/// consolidates updates by writing a full monitor instead. Note that consolidation will
+	/// frequently occur with fewer updates, this number is merely the maximum that may be stored.
+	pub fn new(
+		kv_store: K,
+		logger: L,
+		maximum_pending_updates: u64,
+		entropy_source: ES,
+		signer_provider: SP,
+	) -> Self
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		MonitorUpdatingPersister {
+			kv_store,
+			logger,
+			maximum_pending_updates,
+			entropy_source,
+			signer_provider,
+		}
+	}
+
+	/// Reads channel monitors, along with any stored updates for them.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is extremely important that your [`KVStore::read`] implementation uses the
+	/// [`io::ErrorKind::NotFound`] variant correctly: that is, when a file is not found, and _only_
+	/// in that circumstance (not when there is really a permissions error, for example). This is
+	/// because this function does not list updates. Instead, it lists all monitors, and then using
+	/// their stored `update_id`, synthesizes update storage keys, and tries them in sequence until
+	/// one is not found. All _other_ errors will be bubbled up in this function's [`io::Result`].
+	pub fn read_channel_monitors_with_updates<B: Deref, F: Deref + Clone>(
```

Comment:
If I understand your suggestion correctly, it's that I should extract the work for reading a single monitor and its updates to a function (I think I had it like this once, but I've been trying to inline more). Let's call that function `read_channel_monitor_with_updates`. Then users will deserialize at startup one of two ways:

  - Synchronously, by calling `read_all_channel_monitors_with_updates` (or maybe some shorter name). This function will list the monitors and serially call `read_channel_monitor_with_updates` for each.
  - Asynchronously, by listing channel monitors themselves and spawning tasks that each run `read_channel_monitor_with_updates`, collecting the futures themselves.

In either case they end up with a `Vec<ChannelMonitor>`. 

I think this is good, and I've wondered about something like this myself (my own node is running into startup time issues). I wonder if we should include at least a small example/guidance for how we envision that working, since the function won't be `async` colored. `spawn_blocking`?

Reply:
True, it is parallelizable but not async (though I think you could imagine one wanting to have yield points on reads, I guess that's an exercise for advanced users).

Right, your example shows what I'm getting at. Considerations like limiting the number of simultaneous readers will be important. I don't know where that kind of guidance belongs. Maybe not in LDK at all.

You're right, if we had to populate that HashMap this would have been more annoying to write. Parallelizing would have worked fine with the list-is-truth design, too, since that also had no internal state. I still worry about an edge case or two from not-listing (e.g., a single update write failed silently... you could get penalty txns when we might have been able to throw an error about a gap in the list of updates). But I think it's okay to say people need to be very defensive about storage code in Lightning, and besides, we don't need to relitigate that design choice now.